### DISCUSSION TITLE
Suggested way to handle conflicting modules like opencv and opencv-contrib?

### DISCUSSION THREAD

**Discussion by geroldmeisinger** [Comfy-Org/ComfyUI-Manager#708] [Q&A]
I want to add a [edge drawing preprocessor](https://github.com/Fannovel16/comfyui_controlnet_aux/pull/356) which requires `opencv-contrib-python` . but as the official docu states:

https://pypi.org/project/opencv-python/#installation-and-usage

> There are four different packages (see options 1, 2, 3 and 4 below) and you should SELECT ONLY ONE OF THEM. Do not install multiple different packages in the same environment. There is no plugin architecture: all the packages use the same namespace (cv2). If you installed multiple different packages in the same environment, uninstall them all with pip uninstall and reinstall only one package.

If any non-contrib module is installed I get errors like `AttributeError: module 'cv2.ximgproc' has no attribute 'createEdgeDrawing'`. all packages are installed in the same environment as ComfyUI, so if any custom node down the line installs a non-contrib module, everything breaks.

What is the official way to handle these modules in the ecosystem?

I just stumbled upon the [pip_overrides.json feature](https://github.com/ltdrdata/ComfyUI-Manager/discussions/554) but this is a user settings and cannot be enforced by a custom node (for good reasons, but seems like it should be a default in Comfy-Manager or ComfyUI).

related discussion in official opencv-python: https://github.com/opencv/opencv-python/issues/987#issuecomment-2126354924

**Comment by ltdrdata** [comment]
Oh, this is a tricky issue. This is also connected to this problem.
https://github.com/ltdrdata/ComfyUI-Impact-Pack/pull/266

This is also connected to the issue where it's unclear which opencv package will actually be imported when multiple opencv packages are installed.

The situation I encountered was that both `opencv-python` and `opencv-python-headless` were installed, but their versions were different. When the lower version of opencv was imported, certain APIs were missing.

Since each has different requirements, fixing this to one specific package is inappropriate. To resolve this situation, it seems necessary to allow the user to choose via a prompt or something similar.

As a last resort, you can solve it directly through `install.py`, but this is a very aggressive approach and thus quite challenging.

It is also necessary to consider cases where the dependency on the cv package is indirectly introduced by another package.

This seems to require a discussion.

**Reply by geroldmeisinger** [reply]
it also depends on the order of installation:
`pip install opencv-contrib-python opencv-python` (doesn't work)
is not the same as
`pip install opencv-python opencv-contrib-python` (works)

is it possible to isolate a custom node to it's own environment?
